### PR TITLE
TF-404: Add `.std()` to `Tensor`.

### PR DESCRIPTION
**Pull request by saeta**
This change adds `std()` to compute the standard deviation of a Tensor.
It supports optionally specifying the `alongAxes` parameter, which allows
users to reduce only along specified axes.

Resolves [TF-404](https://bugs.swift.org/browse/TF-404).

### CODE REVIEW

File: stdlib/public/TensorFlow/Ops.swift [hunk 1]
```diff
@@ -577,6 +577,38 @@ public extension Tensor where Scalar : FloatingPoint & Equatable {
...
 
+public extension Tensor where Scalar : FloatingPoint {
+  // TODO: std() should handle non floating point Tensors.
+
+  /// Returns the standard deviation along the specified axes. The reduced
+  /// dimensions are retained with value 1. Does not apply Bessel's correction.
+  /// - Parameter axes: The dimensions to reduce.
+  /// - Precondition: Each value in `axes` must be in the range `-rank..<rank`.
+  @differentiable(wrt: self where Scalar : TensorFlowFloatingPoint)
+  func std() -> Tensor {
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/Ops.swift]: I think I would prefer the full name `func standardDeviation`, despite `std` being a term-of-art and `standardDeviation(alongAxes:)` being really long.

I searched for usages of `std` and I don't think it's a super commonly used operation.

What do you think? @saeta @rxwei

Comment (dan-zheng) [reply stdlib/public/TensorFlow/Ops.swift]: Another reason in favor of `standardDeviation`: abbreviated names like `std` and `stddev` look a bit weird with the `alongAxes:` label.

```swift
x.std()
x.stddev()
x.standardDeviation()

x.std(alongAxes: 0)
x.stddev(alongAxes: 0)
x.standardDeviation(alongAxes: 0)
```

Comment (rxwei) [reply stdlib/public/TensorFlow/Ops.swift]: Completely agreed.

File: stdlib/public/TensorFlow/Ops.swift [hunk 2]
```diff
@@ -577,6 +577,38 @@ public extension Tensor where Scalar : FloatingPoint & Equatable {
   }
 }
 
+public extension Tensor where Scalar : FloatingPoint {
+  // TODO: std() should handle non floating point Tensors.
```
[the remark this answers is not in the document]
Comment (saeta) [reply stdlib/public/TensorFlow/Ops.swift]: Although I'm very tempted to just add the implementation for `Int` backed by `float64`, I'm not quite sure thats always the right answer. (Concretely, float64 is a performance anti-pattern for accelerators.) An alternative would be to make it generic, and require the users to specify the amount of precision they'd like, which could work (at the cost of extra cognitive overhead for users, which we could make go-away in the default case if we can implement defaults for generic type parameters). What do you think about that?

File: stdlib/public/TensorFlow/Ops.swift [hunk 3]
```diff
@@ -577,6 +577,38 @@ public extension Tensor where Scalar : FloatingPoint & Equatable {
...
+
+  /// Returns the standard deviation along the specified axes. The reduced
```
[the remark this answers is not in the document]
Comment (saeta) [reply stdlib/public/TensorFlow/Ops.swift]: Ack, fixed. Thanks!

File: stdlib/public/TensorFlow/Ops.swift (hunk 1 above)
Comment (saeta) [reply stdlib/public/TensorFlow/Ops.swift]: I've now updated the name. Thanks!

**Comment by saeta** [comment]
@swift-ci please test tensorflow

**Comment by saeta** [comment]
CC @jph00 & @sgugger for discussion on the above API design.

**Comment by jph00** [comment]
All reductions should have `squeezingAxes` versions as well (`variance` seems to be missing this too). Users need to be able to choose whether unit axes for the reduced dimensions remain (i.e. for broadcasting) or not.

File: stdlib/public/TensorFlow/Ops.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (dan-zheng) [reply stdlib/public/TensorFlow/Ops.swift]: I think making things generic is a great idea.

How about this?
```swift
extension Tensor where Scalar : Numeric {
  func standardDeviation<FloatingPointScalar>(
    alongAxes axes: Int32..., withScalarType: FloatingPointScalar.Type
  ) where FloatingPointScalar : TensorFlowFloatingPoint { ... }

  // Enabling `standardDeviation(withType: ...)` without specifying axes is important.
  // Perhaps this can be done with a default `[]` axes argument.
  // There may be a better way.
  func standardDeviation<FloatingPointScalar>(
    alongAxes axes: [Int32] = [], withScalarType: FloatingPointScalar.Type
  ) where FloatingPointScalar : TensorFlowFloatingPoint { ... }
}

// Usage:
let x: Tensor<Int32> = ...
x.standardDeviation(withScalarType: Float.self)
x.standardDeviation(alongAxes: 0, withScalarType: Float.self)
```

cc @rxwei

File: stdlib/public/TensorFlow/Ops.swift [hunk 4]
```diff
@@ -577,6 +577,41 @@ public extension Tensor where Scalar : FloatingPoint & Equatable {
   }
 }
 
+public extension Tensor where Scalar : FloatingPoint {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]: Please constrain Scalar on TensorFlowFloatingPoint to be consistent with other extensions. Once that’s done, you should be able to drop all where clauses from the @differentiable attributes you added. This simplified everything.

**Comment by rxwei** [comment]
I agree with Jeremy. Every reduction-like op should have its non-dim-keeping version, aka. ‘squeezingAxes:’.

**Comment by jph00** [comment]
Bias correction should probably be a boolean parameter (also for `variance`).

**Review by rxwei — APPROVED** [review]
LGTM. Feel free to address the ‘squeezingAxes’ variant in a separate patch.

**Comment by sgugger** [comment]
Another comment on consistency: it would be good to have a version of `variance()` with no arguments that takes it over all axes (like `standardDeviation` in this PR, or `mean` and `sum`).

File: stdlib/public/TensorFlow/Ops.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (rxwei) [reply stdlib/public/TensorFlow/Ops.swift]: Your examples above don’t have a return type. Did you mean it’s Tensor<FloatingPointScalar>?

**Comment by dan-zheng** [comment]
As requested, `variance()` and `variance(squeezingAxes:)` added in https://github.com/apple/swift/pull/23811.

cc @jph00 @sgugger: the derivatives for `variance` seem inconsistent with PyTorch.
I'm seeking help, could you please take a look?

File: stdlib/public/TensorFlow/Ops.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (dan-zheng) [reply stdlib/public/TensorFlow/Ops.swift]: > Your examples above don’t have a return type. Did you mean it’s a Tensor of FloatingPointScalar?

Fixed, thanks!

File: stdlib/public/TensorFlow/Ops.swift [hunk 5]
```diff
@@ -578,33 +578,36 @@ public extension Tensor where Scalar : FloatingPoint & Equatable {
...
+  /// Returns the standard deviation of the elements along the specified axes.
+  /// The reduced dimensions are retained with value 1. Does not apply Bessel's
+  /// correction.
   /// - Parameter axes: The dimensions to reduce.
   /// - Precondition: Each value in `axes` must be in the range `-rank..<rank`.
   @differentiable(wrt: self where Scalar : TensorFlowFloatingPoint)
-  func std() -> Tensor {
-    return std(alongAxes: Array(0..<shape.rank))  // Reduce along all dimensions
+  func standardDeviation() -> Tensor {
+    return standardDeviation(alongAxes: Array(0..<shape.rank))  // Reduce along all dimensions
```
[the remark this answers is not in the document]
Comment (saeta) [reply stdlib/public/TensorFlow/Ops.swift]: Fixed (in next push). Good catch. Thanks!

File: stdlib/public/TensorFlow/Ops.swift [hunk 6]
```diff
@@ -577,6 +577,41 @@ public extension Tensor where Scalar : FloatingPoint & Equatable {
   }
 }
 
+public extension Tensor where Scalar : FloatingPoint {
+  // TODO: standardDeviation() should handle non floating point Tensors.
+
+  /// Returns the standard deviation of the elements along the specified axes.
+  /// The reduced dimensions are retained with value 1. Does not apply Bessel's
```
[the remark this answers is not in the document]
Comment (saeta) [reply stdlib/public/TensorFlow/Ops.swift]: Fixed.

File: stdlib/public/TensorFlow/Ops.swift (hunk 4 above)
Comment (saeta) [reply stdlib/public/TensorFlow/Ops.swift]: Thanks done!

File: stdlib/public/TensorFlow/Ops.swift [hunk 7]
```diff
@@ -577,6 +577,42 @@ public extension Tensor where Scalar : FloatingPoint & Equatable {
...
+  // TODO: standardDeviation() should handle non floating point Tensors.
+
+  /// Returns the standard deviation of the elements along the specified axes.
+  /// The reduced dimensions are retained with value `1`. Does not apply
+  /// Bessel's correction.
+  /// - Parameter axes: The dimensions to reduce.
+  /// - Precondition: Each value in `axes` must be in the range `-rank..<rank`.
+  @differentiable(wrt: self)
+  func standardDeviation() -> Tensor {
+    // Reduce along all dimensions
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]: End the comment with a period.

File: stdlib/public/TensorFlow/Ops.swift [hunk 8]
```diff
@@ -577,6 +577,42 @@ public extension Tensor where Scalar : FloatingPoint & Equatable {
...
+  /// - Parameter axes: The dimensions to reduce.
+  /// - Precondition: Each value in `axes` must be in the range `-rank..<rank`.
+  @differentiable(wrt: self)
+  func standardDeviation(alongAxes axes: Int32...) -> Tensor {
+    return standardDeviation(alongAxes: axes)
+  }
+
+  /// Returns the standard deviation of the elements along the specified axes.
+  /// The reduced dimensions are retained with value `1`. Does not apply
+  /// Bessel's correction.
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]: We usually add an empty line between the main doc block and the first parameter item.  Same for other occurrences.

File: test/TensorFlowRuntime/tensor.swift [hunk 9]
```diff
@@ -276,6 +276,26 @@ TensorTests.testAllBackends("SimpleMath") {
                              byError: 0.0001)
 }
 
+TensorTests.testAllBackends("StandardDeviation") {
+  expectEqual(0, Tensor<Float>([1]).standardDeviation().scalarized())
+  expectEqual(
+      0.5,
```
Comment (rxwei) [inline test/TensorFlowRuntime/tensor.swift]: The indentation here should be 2. In the standard library code base, all line breaks are indented by 2.
